05-part_two.Rmd

```diff
@@ -50,23 +111,29 @@ test_that('data dimensions correct', {
 # The second test checks that the maximum value
 # of the variable "some_numbers" does not exceed 1
 test_that('no value greater than 1', {
-  expect_lt(max(mydata$some_numbers), 1)
+  expect_lte(max(mydata$some_numbers), 1)
 })
 
 ```
 
-Expectations are tested using the `test_dir()` function, which takes the path to the testing directory as an argument. 
+Expectations are tested using the `test_dir()` function, which takes the path to the testing directory as an argument:
 ```{r testing, message = FALSE, eval = FALSE, class.source = "badCode"}
-library(testthat)
-library(tidyverse)
-
-# create a data frame
-mydata <- tibble(some_numbers = runif(100))
-
 # test expectations and examine outputs
 # In this case, all tests are OK
 test_dir("testing_directory")
 ```
 
-`test_dir()` provides detailed outputs, including a time line of success/failure and any warnings that may have occured.  
+`test_dir()` provides detailed outputs, including a time line of success/failure and any warnings that may have occurred.
+
+### Testing Shiny apps
+
+R Shiny apps may benefit from a further extended testing approach. They encompass not just some analysis, but also a User Interface and a Server to enable them to run. This makes them closer to a software development project than the average piece of modeling work.
+
+Additional functionality is available within R to support testing of Shiny apps. Further details about this can be found [here](https://shiny.rstudio.com/articles/integration-testing.html).
+
+
+### Test driven development
+
+"Test driven development" describes a way of working in software where tests are written prior to the code being written. This forces developers to sit down and work out what working code encompasses before the write it and avoids any retrofitting of tests to match achieved output. If you have a strong QA plan, you could also implement this for analysis. As testing of each new bit of code happens immediately after creation it also reduced the risk of something faulty getting heavily baked in.
```

Comment:
Do we need 'in software' ?  If we want people to adopt this approach then I think removing that could be beneficial so it doesn't seem like it's only applicable in software dev.

On the 2nd sentence, could we rephrase more positively?  Eg:  "By thinking about what the working code will need to encompass before its written, this helps to avoid retrofitting tests to match achieved output"

Also, typo on 'it also reduceS' (not reduced)